website/project/views/comment.py

```diff
@@ -188,40 +285,79 @@ def is_reply(target):
 @must_be_contributor_or_public
 def list_comments(auth, node, **kwargs):
     anonymous = has_anonymous_link(node, auth)
+    page = request.args.get('page')
     guid = request.args.get('target')
-    target = resolve_target(node, guid)
-    serialized_comments = serialize_comments(target, auth, anonymous)
-    n_unread = 0
+    root_id = request.args.get('rootId')
+    is_list = request.args.get('isCommentList')
+    is_widget = False
+
+    if page == 'total' and not root_id:  # "Total" on discussion page
+        comments = list_total_comments(node, auth, 'total')
+    elif page == 'total':  # Discussion widget on overview's page
+        comments = list_total_comments_widget(node, auth)
+        is_widget = True
+    elif not root_id:  # Overview/Files/Wiki page on discussion page
+        comments = list_total_comments(node, auth, page)
+    else:
+        target = resolve_target(node, page, guid)
+        comments = getattr(target, 'commented', [])
 
+    n_unread = 0
     if node.is_contributor(auth.user):
         if auth.user.comments_viewed_timestamp is None:
             auth.user.comments_viewed_timestamp = {}
             auth.user.save()
-        n_unread = n_unread_comments(target, auth.user)
-    return {
-        'comments': serialized_comments,
+        n_unread = n_unread_comments(node, auth.user, page, root_id)
+
+    ret = {
+        'comments': [
+            serialize_comment(comment, auth, anonymous)
+            for comment in comments
+        ],
         'nUnread': n_unread
     }
+    if is_list:
+        discussions = comment_discussion(comments, node, anonymous=anonymous, widget=is_widget)
+        ret.update(discussions)
 
+    return ret
 
-def n_unread_comments(node, user):
-    """Return the number of unread comments on a node for a user."""
-    default_timestamp = datetime(1970, 1, 1, 12, 0, 0)
-    view_timestamp = user.comments_viewed_timestamp.get(node._id, default_timestamp)
-    return Comment.find(Q('node', 'eq', node) &
-                        Q('user', 'ne', user) &
-                        Q('date_created', 'gt', view_timestamp) &
-                        Q('date_modified', 'gt', view_timestamp)).count()
+
+def list_total_comments_widget(node, auth):
+    comments = list(Comment.find(Q('node', 'eq', node)))
+    comments.sort(
+        key=lambda item: item.date_created,
+        reverse=False
+    )
+    return comments
+
+
+def list_total_comments(node, auth, page):
+    if page == 'total':
+        comments = list(Comment.find(Q('node', 'eq', node)))
+    else:
+        comments = list(Comment.find(Q('node', 'eq', node) &
+                                Q('page', 'eq', page)))
+
+    root_comments = []
+    for comment in comments:
+        if not isinstance(comment.target, Comment):
+            root_comments.append(comment)
+    root_comments = sorted(
+        root_comments,
+        key=lambda item: item.date_created,
```

Comment:
Again, sorting should be done in the database when possible. Modular-odm supports sorting with MongoDB.